LLA to NED abstraction

erika-tsutsumi opened the pull request:
As per issue listed below, introduces functions to generate LLA to NED conversion factors as well as a generalized 3d coordinate transformation function that can be used to do the conversion. Also contains unit tests for all functions introduced. 

Fixes https://github.com/TauLabs/TauLabs/issues/483

erika-tsutsumi committed:
CoordinateConversions: functions to linearize NED to LLA conversion

Takes home latitude and altitude to get conversion factor(in degrees/meter)
between north-east and longitude-latitude.

Separate functions exist for both float and double inputs.

erika-tsutsumi committed:
Unit Tests: test accuracy of LLA_to_NED linearization factors

Calculate LLA_to_NED conversion factors using LLA_to_NED abstraction
function and compare with calculated conversion factor (from matlab). Uses
three sample locations (Equator, Boston USA, Sydney AUS). Function passes
all 3 tests.

erika-tsutsumi committed:
Unit Tests: Test LLA_to_NED linearization factor generation and usage

1) Unit tests for LLA_linearization_float and LLA_linearization_double
functions which check linearization factors against calculated values
(using the same formula in python). Each function is tested for several
test locations in different hemispheres and at different elevations.

2) Tests the get_linearized_3D_transformation function which takes the
linearization factors as an input to get the NED coordinates from an LLA
coordinate and home location.

kubark42 commented:
Jenkins test this please.

kubark42 commented on flight/Libraries/math/coordinate_conversions.c:
```diff
@@ -366,6 +366,75 @@ void rot_mult(float R[3][3], const float vec[3], float vec_out[3], bool transpos
+void LLA_linearization_float(int32_t home_latitude_1e7, float home_altitude, float T[])
+{
+    float lat_rad = home_latitude_1e7 / 10.0e6f * DEG2RAD;
+	float alt = home_altitude;
```
Whitespace here and below.

kubark42 commented on flight/Modules/Attitude/attitude.c:
```diff
@@ -154,6 +154,8 @@ static const float zeros[3] = {0.0f, 0.0f, 0.0f};
 static struct complementary_filter_state complementary_filter_state;
 static struct cfvert cfvert; //!< State information for vertical filter
 
+static float T[3];
```
As long as you're moving `T` around, you might consider renaming the transformation with a more searchable, i.e. distinct, name.

kubark42 commented on flight/Modules/CameraStab/camerastab.c:
```diff
@@ -58,6 +59,7 @@
 #include "positionactual.h"
 #include "tabletinfo.h"
 
+
```
Was this whitespace intentional?

kubark42 commented on flight/Modules/Sensors/simulated/sensors.c:
```diff
@@ -478,8 +476,8 @@ static void simulateModelQuadcopter()
-		gpsPosition.Latitude = homeLocation.Latitude + ((pos[0] + gps_drift[0]) / T[0] * 10.0e6);
-		gpsPosition.Longitude = homeLocation.Longitude + ((pos[1] + gps_drift[1])/ T[1] * 10.0e6);
+		gpsPosition.Latitude = homeLocation.Latitude + ((pos[0] + gps_drift[0]) / T[0]);
+		gpsPosition.Longitude = homeLocation.Longitude + ((pos[1] + gps_drift[1])/ T[1]);
```
Could you fix the whitespace around the `/`?

kubark42 commented on flight/Modules/Sensors/simulated/sensors.c:
```diff
@@ -779,8 +775,8 @@ static void simulateModelAirplane()
-		gpsPosition.Latitude = homeLocation.Latitude + ((pos[0] + gps_drift[0]) / T[0] * 10.0e6);
-		gpsPosition.Longitude = homeLocation.Longitude + ((pos[1] + gps_drift[1])/ T[1] * 10.0e6);
+		gpsPosition.Latitude = homeLocation.Latitude + ((pos[0] + gps_drift[0]) / T[0]);
+		gpsPosition.Longitude = homeLocation.Longitude + ((pos[1] + gps_drift[1])/ T[1]);
```
Likewise.

kubark42 commented on flight/Modules/Sensors/simulated/sensors.c:
```diff
@@ -1036,8 +1030,8 @@ static void simulateModelCar()
-		gpsPosition.Latitude = homeLocation.Latitude + ((pos[0] + gps_drift[0]) / T[0] * 10.0e6);
-		gpsPosition.Longitude = homeLocation.Longitude + ((pos[1] + gps_drift[1])/ T[1] * 10.0e6);
+		gpsPosition.Latitude = homeLocation.Latitude + ((pos[0] + gps_drift[0]) / T[0]);
+		gpsPosition.Longitude = homeLocation.Longitude + ((pos[1] + gps_drift[1])/ T[1]);
```
One last whitespace around `/`.

kubark42 commented on flight/tests/coordinate_conversions/unittest.cpp:
```diff
@@ -85,3 +85,306 @@ TEST_F(RneFromLLATest, Equator) {
+  LLA_linearization_float(homeLLA_D[0], homeLLA_D[2], T);
+
+  float eps = 0.000001f;
+
```
Probably should add a comment that you're looking for the ratio between expected and returned values, and not absolute differences.

kubark42 commented on flight/tests/coordinate_conversions/unittest.cpp:
```diff
@@ -85,3 +85,306 @@ TEST_F(RneFromLLATest, Equator) {
+TEST_F(LLALineraization, Equator_float) {
+  // Test location is Equator
+  float homeLLA_D[] = {0, 0, 0};
+  float dN = 0.0111319490793274;
```
It would be helpful to embed the python script used for generating these values.

kubark42 commented on flight/tests/coordinate_conversions/unittest.cpp:
```diff
@@ -85,3 +85,306 @@ TEST_F(RneFromLLATest, Equator) {
+                                   homeLLA_D[0], homeLLA_D[1], homeLLA_D[2],
+                                   T, NED);
+
+  // 0.0110574275839048
```
Remove commented code.

kubark42 commented on flight/tests/coordinate_conversions/unittest.cpp:
```diff
@@ -85,3 +85,306 @@ TEST_F(RneFromLLATest, Equator) {
+
+  float eps = 0.003f;
+
+  double N_expected = 0.01110805;
```
Likewise, embedding the python code would be good here for repeatability.

kubark42 commented on flight/tests/coordinate_conversions/unittest.cpp:
```diff
@@ -85,3 +85,306 @@ TEST_F(RneFromLLATest, Equator) {
+                                   homeLLA_D[0], homeLLA_D[1], homeLLA_D[2],
+                                   T, NED);
+
+  float eps = 0.003f;
```
So I guess this means that the linearization results in values that are only correct to 99.7%?

kubark42 commented:
jenkins: add to whitelist

erika-tsutsumi committed:
CoordinateConversions: introduce abstract function to map LLA to NED

Introduce function to convert LLA position data (from GPS or
other source) to NED. Uses LLA to NED conversion factors from
LLA_linearization_float() or LLA_linearization_double().

The conversion function is abstracted to be usable for any coordinate
system conversion/mapping provided the input conversion factors are
correct. Note that inputs coordinates are oddly typed (int32_t, int32_t, float)
because this originated as a function which took LLA * 1e7 position data.

Factor out duplicated code throughout project.

kubark42 commented on flight/Modules/Sensors/simulated/sensors.c:
```diff
@@ -467,10 +467,8 @@ static void simulateModelQuadcopter()
-		T[1] = cosf(homeLocation.Latitude / 10e6 * DEG2RAD)*(homeLocation.Altitude+6.378137E6) * DEG2RAD;
-		T[2] = -1.0;
-		
+		LLA_linearization_double(home_latitude, home_altitude, T);
```
This needs to be changed to `homeLocation.Latitude, homeLocation.Altitude`.

erika-tsutsumi committed:
Fixed whitespace issues requested in review

kubark42 commented on flight/Modules/Attitude/attitude.c:
```diff
@@ -1051,10 +1056,10 @@ static int32_t updateAttitudeINSGPS(bool first_run, bool outdoor_mode)
 	mag_updated &= !outdoor_mode || (homeLocation.Be[0] != 0 || homeLocation.Be[1] != 0 || homeLocation.Be[2]);
 
-	// A more stringent requirement for GPS to initialize the filter
+        // A more stringent requirement for GPS to initialize the filter
```
whitespace

kubark42 commented on flight/Modules/Attitude/attitude.c:
```diff
@@ -1068,10 +1073,10 @@ static int32_t updateAttitudeINSGPS(bool first_run, bool outdoor_mode)
 			set_state_estimation_error(SYSTEMALARMS_STATEESTIMATION_UNDEFINED);
 
-	} else if (outdoor_mode && (gpsData.Satellites < insSettings.MinRNAVSatellites || gpsData.PDOP > insSettings.MinRNAVPDOP)) {
-		if (gpsData.Satellites < insSettings.MinRNAVSatellites)
```
whitespace

kubark42 commented on flight/Modules/ManualControl/tablet_control.c:
```diff
@@ -207,13 +208,8 @@ static int32_t tabletInfo_to_ned(TabletInfoData *tabletInfo, float *NED)
-	T[0] = alt+6.378137E6f;
-	T[1] = cosf(lat)*(alt+6.378137E6f);
-	T[2] = -1.0f;
+	LLA_linearization_float(homeLocation.Latitude, homeLocation.Altitude, T);
```
This shouldn't be done every loop. It's not a blocker for this PR, but if you wanted to you could move it out.

kubark42 commented:
I like this a lot. It's very nice how you folded the entire linearization into the matrix `T`, including the shifts by `1e7`. The resulting code is more efficient in RAM and CPU both, which is a big win.

@peabody124 I'll test fly it, but I don't have a camera rig to make sure the CameraStab module still works correctly. Is this something you can take a look at?

peabody124 commented:
I'll review the code shortly. No guarantees on flying with my camera rig, but I'll try.

kubark42 commented:
Cool. 

I was looking in more detail and realized that this generalized-linearization lets us move to ECEF-based navigation. In theory, this will be a a nice precision and latency win. Preparing a PR now.

kubark42 commented on flight/Libraries/math/coordinate_conversions.c:
```diff
@@ -366,6 +366,75 @@ void rot_mult(float R[3][3], const float vec[3], float vec_out[3], bool transpos
 }
 
 /**
+ * @brief LLA_linearization_float calculates linearized NED to LLA conversion using floats
```
Is this comment right? Isn't it the other way around, i.e. this is performing an LLA-->NED transformation linearization?

erika-tsutsumi replied:
Yup. Fixed.

kubark42 commented on flight/Libraries/math/coordinate_conversions.c:
```diff
@@ -366,6 +366,75 @@ void rot_mult(float R[3][3], const float vec[3], float vec_out[3], bool transpos
+ * @param home_altitude
+ * @param T is the linearized NED to LLA conversion
+ */
+void LLA_linearization_float(int32_t home_latitude_1e7, float home_altitude, float T[])
```
Might be called `LLA2NED_linearization...`. This way it is more explicit about what the operation is, and it makes space for my `ECEF2NED_linearization...`.

erika-tsutsumi replied:
Also fixed.

kubark42 commented on flight/Libraries/math/coordinate_conversions.c:
```diff
@@ -366,6 +366,75 @@ void rot_mult(float R[3][3], const float vec[3], float vec_out[3], bool transpos
+ * @param [in] C mapping coefficient
+ * @param [out] XYZ output coordinates
+ */
+void get_linearized_3D_transformation(const int32_t theta_prime, const int32_t phi_prime, const float r_prime,
```
@peabody124 do you have any thoughts on making this more general? It bothers me that there is a mix of  `int32_t`  and `float` arguments. I see why @erika-tsutsumi did it this way, but it's not as clean as I'd like.

One thing we could do is turn `GPSPosition.Altitude` into a `int32_t`, i.e. remove the `GpsPosition->Altitude = (float)posllh->hMSL*0.001f` conversion in the GPS module. Then this function could be called completely on `int32_t`, which should be more maintainable.

Thoughts?

P.S. @erika-tsutsumi there's no need to change now, this would be for a future PR.

peabody124 replied:
I don't have a strong opinion on the units. Given we always use int32_t throughout the code this is the most efficient way to do it and probably makes sense. Casting the int32 to float directly would lose precision, and we don't want to add the burden of using doubles.

kubark42 replied:
https://github.com/TauLabs/TauLabs/blob/next/flight/Modules/GPS/UBX.c#L206 is what I was thinking of getting rid of. The native input to the system is in [mm] so I'm proposing we could remove the conversion to `float` and then the rest follows naturally. Was that clear?

On the other hand, we've been playing around with RTK GPS and sub-mm level precisioning is already something we're having fun with. So it might not make a ton of sense to optimize this for older tech GPSes when we really should be aiming at supporting an order of magnitude more precision in the near future.

kubark42 commented on flight/Modules/Attitude/attitude.c:
```diff
@@ -827,7 +828,11 @@ static int32_t setNavigationRaw()
+		get_linearized_3D_transformation(gpsPosition.Latitude,  gpsPosition.Longitude, gpsPosition.Altitude,
+		                                 homeLocation.Latitude, homeLocation.Longitude,  homeLocation.Altitude,
+		                                 T, NED);
+
```
Extra newline.

kubark42 commented:
@erika-tsutsumi there's also a hidden conversion at https://github.com/TauLabs/TauLabs/blob/54fff14e8b863ba7d106b8c208dfec2cc380c3b8/flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c#L259. This should be cleaned up as well.

peabody124 commented:
Code committed looks good. @erika-tsutsumi agreed it would be great to get the UAVOMSPBridge changed as well for consistency. Can you comment if you will be able to do this? Kenn has a few small comments that would be nice to fix, none major or blockers.

Has this been tested outside?

peabody124 commented:
Tested in flight simulation without issue.

peabody124 commented:
Simulation also shows the camera tracking working well.

erika-tsutsumi commented:
@peabody124 yup I'll look at the UAVOMSPBridge  when I have a chance

erika-tsutsumi committed:
CoordinateConversions: rename variable T to searchable name

erika-tsutsumi committed:
CoordinateConversions: refactor LLA2NED conversions + add instance

1) Change variable T to linearized_conversion_factor_f or
linearized_conversion_factor_d for searchability

2) Changed name of LLA_linearization_float and LLA_linearization_double
to LLA2NED_... for clarity

3) Fixed whitespace issues

4) Added one implementation of the LLA to NED conversion function in
UAVOMSPBridge

erika-tsutsumi commented:
For the UAVOMSPBridge, the math wasn't quite the same so could someone double check it? I left all the original code in comments because I wasn't sure about it, I'll remove them once someone else okays it.

peabody124 commented on flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c:
```diff
@@ -256,20 +257,31 @@ static void msp_send_comp_gps(struct msp_bridge *m)
+			float deltas[3];
+			get_linearized_3D_transformation(gps_data.Latitude, gps_data.Longitude, gps_data.Altitude,
+			                                 home_data.Latitude, home_data.Longitude, home_data.Altitude,
+			                                 linearized_conversion_factor[], deltas[]);
```
`linearized_conversion_factor` dropped the `_f` and also the [] should not be there

erika-tsutsumi replied:
yup sorry I was building flyingf3 and so didn't catch it, should be updated now

peabody124 commented on flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c:
```diff
@@ -256,20 +257,31 @@ static void msp_send_comp_gps(struct msp_bridge *m)
-			if ((delta_lon == 0) && (delta_lat == 0))
-				data.comp_gps.direction_to_home = 0;
+//			data.comp_gps.distance_to_home  = (uint16_t)(sqrtf(delta_x * delta_x + delta_y * delta_y) * 1e-4f);  // meters
+			data.comp_gps.distance_to_home  = (uint16_t)(sqrtf(deltas[0] * deltas [0] + deltas[1] * deltas[1]));  // meters
```
I'm trying to track down why the difference of 1e-4f

kubark42 replied:
Took me a second, but I think I figured it out. The comments like `float delta_y = delta_lon * WGS84_RADIUS_EARTH_KM * DEG2RAD;  // KM * 1e7` point out that the unit is already `1e7`. So `1e-4` is appropriate for converting the final output to meters.

peabody124 replied:
Ok, great. So math is good?

peabody124 replied:
If so, looks ok to merge to me once comments removed.

kubark42 commented on flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c:
```diff
@@ -256,20 +257,31 @@ static void msp_send_comp_gps(struct msp_bridge *m)
+//			delta_y *= cosf(home_data.Latitude * 1e-7f * (float)DEG2RAD);  // Latitude compression correction
+
+			float linearized_conversion_factor_f[3];
+			LLA2NED_linearization_float(home_data.Latitude, home_data.Altitude, linearized_conversion_factor_f);
```
I'm unsure about how I feel about doing the linearization each loop. @peabody124, what's your opinion about increased code complexity vs. the CPU cost of doing this per-loop? I guess I'm kind of okay giving it a pass, since it's just a few Hz.

erika-tsutsumi commented:
Rebased commits in preparation for merge.

peabody124 commented:
LGTM @kubark42 merge away if good with you

P.S. i'm ok with a few loops calling it extra times for code simplicity. we can change that later.

kubark42 commented on flight/Modules/CameraStab/camerastab.c:
```diff
@@ -376,22 +377,18 @@ static void tablet_info_process() {
-		lat = homeLocation.Latitude / 10.0e6f * DEG2RAD;
-		alt = homeLocation.Altitude;
+		float linearized_conversion_factor_f[3];
+		LLA2NED_linearization_float(homeLocation.Latitude, homeLocation.Altitude, linearized_conversion_factor_f);
```
In a future PR, this should be moved out of the camera stab loop. No sense in recalculating it every time.

kubark42 commented:
LGTM and tests fine with a real GPS. Merging.

@peabody124, maybe we should consider moving this linearization into a UAVObject. It seems like it's used by several modules and I foresee it being used by more in the future. Might make sense to think about sharing instead of each module having its own private instance.

peabody124 commented:
@kubark42 that seems like a great idea and gets rid of all the duplicate linearizations!
